### PR TITLE
tests: use apt as compatible with trusty

### PR DESCRIPTION
**Pull request by vosst**

### CODE REVIEW

**Comment by zyga** [comment]
Looks good to me

File: spread.yaml [hunk 1]
```diff
@@ -68,22 +68,22 @@ prepare: |
...
 
     # needed so that we have golang-gopkg-macaroon.v1 which is not (yet)
     # in trusty
     add-apt-repository --update ppa:snappy-dev/image
     # this should not be needed but apparently it is :/
-    apt update
+    apt-get update
 
-    apt build-dep -y ./
+    apt-get install -y $(gdebi --quiet --apt-line ./debian/control)
```
Comment (chipaca) [inline spread.yaml]: why this? trusty had `apt-get build-dep`

File: tests/upgrade/basic/task.yaml [hunk 2]
```diff
@@ -18,7 +18,7 @@ execute: |
     echo Do upgrade
     # allow-downgrades prevents errors when new versions hit the archive, for instance,
     # trying to install 2.11ubuntu1 over 2.11+0.16.04
-    apt install -y --allow-downgrades ${SPREAD_PATH}/../snapd*.deb
+    dpkg -i ${SPREAD_PATH}/../snapd*.deb || apt-get -f install -y
```
Comment (chipaca) [inline tests/upgrade/basic/task.yaml]: does this do what we want when the .deb is nominally older than the one available to apt-get?

File: spread.yaml (hunk 1 above)
Comment (chipaca) [inline spread.yaml]: aaand i just realised it's using the local thing.

File: tests/upgrade/basic/task.yaml (hunk 2 above)
Comment (vosst) [inline tests/upgrade/basic/task.yaml]: yup --force-downgrade is selected by default by dpkg.

Comment (chipaca) [inline tests/upgrade/basic/task.yaml]: I mean, won't apt-get then decide the easiest way to fix any issue is to re-upgrade snapd?

Comment (vosst) [inline tests/upgrade/basic/task.yaml]: I think @mvo5 can shed a little more light here.

File: tests/lib/prepare.sh [hunk 3]
```diff
@@ -3,7 +3,7 @@
 set -eux
 
 prepare_classic() {
-    apt install -y ${SPREAD_PATH}/../snapd_*.deb
+    dpkg -i ${SPREAD_PATH}/../snapd_*.deb || apt-get -f install -y
```
Comment (pedronis) [inline tests/lib/prepare.sh]: this seems a candidate for a small shell function too

Comment (vosst) [inline tests/lib/prepare.sh]: I actually introduced a small shell function before. Turns out the one-liner looks better and serves well as a reminder that trusty is the baseline in terms of features. I'm happy to comply if you insist on a function, though :)

Comment (pedronis) [inline tests/lib/prepare.sh]: I won't insist, it's a bit borderline, thanks for doing the functions for systemd-run though

Comment (vosst) [inline tests/lib/prepare.sh]: ack and thx

File: tests/lib/prepare.sh [hunk 4]
```diff
@@ -3,7 +3,8 @@
 set -eux
 
 prepare_classic() {
-    apt install -y ${SPREAD_PATH}/../snapd_*.deb
+    # relying on dpkg as apt(-get) does not support installation from local files in trusty.
+    dpkg -i ${SPREAD_PATH}/../snapd_*.deb || apt-get -f install -y
```
Comment (niemeyer) [inline tests/lib/prepare.sh]: As commented on another PR, let's please not do this as it's guessing the reason for the failure and introducing an alternative path regardless. It's much more clear to act explicitly for reasons we know about. For example:
```
if [ "$SPREAD_SYSTEM" = "ubuntu-14.04-64" ]; then 
    ...
else
    ...
fi
```

Comment (niemeyer) [inline tests/lib/prepare.sh]: There are other instances where this is valid below.

Comment (vosst) [inline tests/lib/prepare.sh]: Yup, I think this change would benefit from an apt helper function taking care of conditionally executing the right cmd then. What do you think?

File: tests/lib/prepare.sh (hunk 3 above)
Comment (vosst) [inline tests/lib/prepare.sh]: Introduced a small helper "apt_install_local_packages"

File: tests/lib/prepare.sh (hunk 4 above)
Comment (vosst) [inline tests/lib/prepare.sh]: I introduced a helper conditionally calling the right cmds based on SPREAD_SYSTEM, also adjusted CUPS package installation/remove taking SPREAD_SYSTEM into account.

File: spread.yaml (hunk 1 above)
Comment (pedronis) [inline spread.yaml]: if I understand @mvo5 was proposing to leave the future way at least as a comment

# in 16.04: apt build-dep -y ./

File: tests/lib/apt.sh [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+apt_install_local_packages() {
+    if [ "$SPREAD_SYSTEM" = "ubuntu-14.04-64" ]; then
+        # relying on dpkg as apt(-get) does not support installation from local files in trusty.
+        dpkg -i "$@" || apt-get -f install -y
```
Comment (niemeyer) [inline tests/lib/apt.sh]: Again, let's please drop the || here. This logic will ignore real problems. If the purpose is to ignore dependencies, let's use the respective flag of dpkg, and call apt-get on the next line.

File: tests/lib/apt.sh [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+apt_install_local_packages() {
```
Comment (niemeyer) [inline tests/lib/apt.sh]: s/_packages// please.. it could only possibly be a package.

File: tests/upgrade/basic/task.yaml [hunk 7]
```diff
@@ -3,8 +3,11 @@ systems: [-ubuntu-core-16-64]
 restore: |
     rm -f /var/tmp/myevil.txt
 execute: |
+    # pull in apt helpers
```
Comment (niemeyer) [inline tests/upgrade/basic/task.yaml]: I'd suggest dropping these comments (here and in the other instance above). The line below feels super clear.

**Comment by niemeyer** [comment]
@vosst Tests still failing, not all points covered yet.

File: tests/upgrade/basic/task.yaml (hunk 7 above)
Comment (vosst) [inline tests/upgrade/basic/task.yaml]: Fixed.

File: tests/lib/apt.sh (hunk 5 above)
Comment (vosst) [inline tests/lib/apt.sh]: Fixed.

File: tests/lib/apt.sh (hunk 6 above)
Comment (vosst) [inline tests/lib/apt.sh]: Fixed.

File: spread.yaml (hunk 1 above)
Comment (vosst) [inline spread.yaml]: Fixed.

File: tests/lib/prepare.sh [hunk 8]
```diff
@@ -38,8 +38,8 @@ prepare_classic() {
 
 setup_reflash_magic() {
         # install the stuff we need
-        apt install -y kpartx busybox-static
-        apt install -y ${SPREAD_PATH}/../snapd_*.deb
+        apt-get install -y kpartx busybox-static
+        dpkg -i ${SPREAD_PATH}/../snapd_*.deb || apt-get -f install -y
```
Comment (vosst) [inline tests/lib/prepare.sh]: Introduced a helper.

**Comment by vosst** [comment]
Previous spread test failure seems to be spurious, with interfaces-cups-control timing out when downloading packages. Pushed update for documentation request.

**Pull request merged**